rasa_nlu_examples/featurizers/sparse/hashing_featurizer.py

```diff
@@ -0,0 +1,139 @@
+from typing import Any, Optional, Text, Dict, List, Type
+
+import numpy as np
+from sklearn.feature_extraction.text import HashingVectorizer
+
+from rasa.nlu.components import Component
+from rasa.nlu.featurizers.featurizer import SparseFeaturizer
+from rasa.nlu.model import Metadata
+from rasa.nlu.config import RasaNLUModelConfig
+from rasa.nlu.tokenizers.tokenizer import Tokenizer
+from rasa.shared.nlu.constants import TEXT, FEATURE_TYPE_SENTENCE, FEATURE_TYPE_SEQUENCE
+from rasa.shared.nlu.training_data.message import Message
+from rasa.shared.nlu.training_data.features import Features
+from rasa.shared.nlu.training_data.training_data import TrainingData
+from rasa.nlu.constants import (
+    DENSE_FEATURIZABLE_ATTRIBUTES,
+    FEATURIZER_CLASS_ALIAS,
+    TOKENS_NAMES,
+)
+
+
+class HashingFeaturizer(SparseFeaturizer):
+    """
+    This component adds sparse features by applying a hashing vectorizer to the input tokens. Each token gets mapped
+    to an index in a sparse vector that is computed by applying a MurmurHash3 function and wrapping the resulting
+    integer to the number of columns in the vector.
+
+    The component is stateless and does not require training. For a small number of features, hash collisions get
+    more likely, meaning that multiple words get mapped to the same index. It is not possible to compute the inverse
+    transform to determine which tokens belong to a specific vector representation.
+
+    The implementation used is Scikit-Learn's HashingVectorizer class, which is described in detail here:
+    https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.HashingVectorizer.html
+    """
+
+    @classmethod
+    def required_components(cls) -> List[Type[Component]]:
+        return [Tokenizer]
+
+    @classmethod
+    def required_packages(cls) -> List[Text]:
+        return ["sklearn"]
+
+    defaults = {
+        # The following parameters and defaults are similar as the ones used by
+        # scikit-learn (version 0.24.2). For detailed explanations on these parameters
+        # and their defaults, have a look at the scikit-learn docs:
+        # https://scikit-learn.org/0.24/modules/generated/sklearn.feature_extraction.text.HashingVectorizer.html
+        #
+        # the number of columns in the output vector
+        "n_features": 2 ** 20,  # int
```

Comment:
I'd say it's fine to say this to 2000 for now. I think our DIET won't be able to handle a vocabulary this size. I could be wrong, but I believe it turns it all into a dense vector at some point.